src/inferencer/constraintStore.ts

```diff
@@ -48,12 +56,19 @@ function solveConstraint(constraintLhs: Type, constraintRhs: Type): any | undefi
     return
   }
   // Rule 4
-  // else if (isTypeVariable(constraintLhs) && is_function_type(constraintStore.get(constraintRhs))
-  //   && ) {
-  //   // error
-  //   console.log('[debug] Error in Rule 4!')
-  //   return {constraintLhs: constraintLhs, constraintRhs: constraintRhs} // for error logging
-  // }
+  else if (
+    (isTypeVariable(constraintLhs) &&
+      constraintStore.get(constraintRhs) !== undefined &&
+      isFunctionType(constraintStore.get(constraintRhs)) &&
+      (constraintStore.get(constraintRhs) as FunctionType).parameterTypes.includes(
```

Comment:
I don't think the `includes` work here. The parameterTypes array contains objects. Comparison of objects is based on reference rather than what's inside the object.